Add arbitrary tag capability to protobuf output in dnsdist

sethomisc opened the pull request:
### Short description
<!-- Write a small description of what this Pull Request fixes or provides, including the issue #s -->
New Lua commands for storing text table information in DNSQuestion and reading the table in DNSResponse.  Additionally storing table in DNSProtoBufMessage as tags along with a function to alter the protobuf message from a 'query' to that of a 'response'.   These added functions are to allow reporting of Lua based RPZ detection and reporting. Notes, example scripts and config file are located in the directory pdns/zzz-gca-example  Tests will be coming soon.
### Checklist
<!-- please indicate if any of these things are done/included with this Pull Request. Not all boxes need to be checked for the Pull Request to be accepted -->
I have:
- [ X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [ X] compiled and tested this code
- [ X] included documentation (including possible behaviour changes)
- [ X] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)
- [X ] 
[dnsdist-Lua additions-June_7_2017.pdf](https://github.com/PowerDNS/pdns/files/1072420/dnsdist-Lua.additions-June_7_2017.pdf)

johnhtodd commented:
A better title for this PR would be "Add arbitrary tag capability to protobuf output in dnsdist"

Habbie commented:
> A better title for this PR would be "Add arbitrary tag capability to protobuf output in dnsdist"

Title updated!

rgacogne commented on pdns/dnsdist.hh:
```diff
@@ -51,6 +59,123 @@ extern uint16_t g_ECSSourcePrefixV4;
+// ----------------------------------------------------------------------------
+// getEntry() - return the specified tag entry
+// ----------------------------------------------------------------------------
+std::string getEntry(int iEntry) const
```
Please don't use a signed type for an unsigned value, `size_t` or at least `unsigned int` would be more appropriate.

sethomisc replied:
changed to used size_t

rgacogne commented on pdns/dnsdist.hh:
```diff
@@ -51,6 +59,123 @@ extern uint16_t g_ECSSourcePrefixV4;
+std::string getMatch(const std::string& strLabel)  const
+{
+
+    std::unordered_map<std::string, std::string>::const_iterator got =tagData.find (strLabel);
```
Looks like a good candidate for using `auto` instead of the complete type.

sethomisc replied:
Changed to use auto.

rgacogne commented on pdns/dnsdist.hh:
```diff
@@ -51,6 +59,123 @@ extern uint16_t g_ECSSourcePrefixV4;
+int iCounter = 0;
+
+   std::unordered_map<std::string, std::string>::const_iterator itr;
+   for (itr =tagData.begin(); itr != tagData.end(); itr++)
```
Perhaps `for (const auto& itr : tagData) {` would be easier to read?

sethomisc replied:
Changed to use auto to be easier to read.

rgacogne commented on pdns/dnsdist.hh:
```diff
@@ -51,6 +59,123 @@ extern uint16_t g_ECSSourcePrefixV4;
+// ----------------------------------------------------------------------------
+// count() - return number of tag entries
+// ----------------------------------------------------------------------------
+int count() const
```
`size_t` or `unsigned int`

sethomisc replied:
Replaced with size_t

rgacogne commented on pdns/dnsdist.hh:
```diff
@@ -71,6 +196,7 @@ struct DNSQuestion
   bool skipCache{false};
   bool ecsOverride;
   bool useECS{true};    
+  QTag qTag;                        // GCA - Seth Ornstein - extra class for tags 5/30/2017
```
Since we are trying to keep the `DNSQuestion` object as light as possible for performance reason, would you be willing to consider making that an `std::unique_ptr<QTag>` so that we don't pay the cost of constructing the object until we really use it? Please also move it up in the `DNSQuestion` object with the other pointers to make padding more optimal.
By the way, if I'm not mistaken, the `QTag` object is basically a `std::unordered_map<std::string,std::string>`, perhaps we could avoid defining a new object and just handle the needful in the `Lua` bindings?

rgacogne commented on pdns/protobuf.cc:
```diff
@@ -97,6 +97,57 @@ void DNSProtoBufMessage::setEDNSSubnet(const Netmask& subnet, uint8_t mask)
+// ----------------------------------------------------------------------------
+
+void DNSProtoBufMessage::addTags(const std::string& strLabel, const std::string& strValue)
+{
```
I think we should move `RecProtoBufMessage::setPolicyTags()` and `RecProtoBufMessage::getPolicyTags()` here instead.

rgacogne commented on pdns/protobuf.cc:
```diff
@@ -97,6 +97,57 @@ void DNSProtoBufMessage::setEDNSSubnet(const Netmask& subnet, uint8_t mask)
+        cTemp[2] = 0;
+        cTemp[3] = 1;
+        rr->set_rdata(cTemp, 4);
+      }
```
This method only adds one RR, it should probably called `addRR()` instead. should Please don't hardcode the type, class, TTL and content here, those should be parameters passed to this method.

rgacogne commented on pdns/dnsdist-lua.cc:
```diff
@@ -1572,6 +1572,48 @@ vector<std::function<void(void)>> setupLua(bool client, const std::string& confi
+     g_lua.registerFunction<string(DNSQuestion::*)(std::string)>("getTagMatch", [](const DNSQuestion& dq, const std::string& strLabel) {
+
+        std::string strValue = dq.qTag.getMatch(strLabel);
+        return(strValue);
```
The parentheses are not needed for this kind of `return` statement.

sethomisc replied:
Parentheses removed for the return.

sethomisc commented on pdns/dnsdist-lua.cc:
```diff
@@ -1572,6 +1572,48 @@ vector<std::function<void(void)>> setupLua(bool client, const std::string& confi
+
+  // --------------------------------------------------------------------------
+  // GCA - Seth Ornstein added lua callable functions - 6/2/2017
+  // DNSQuestion - setTag, setTagArray, getTagMatch, getTagArray
```
Comments removed.

sethomisc replied:
Comments removed.

sethomisc commented on pdns/dnsdist-lua.cc:
```diff
@@ -1572,6 +1572,48 @@ vector<std::function<void(void)>> setupLua(bool client, const std::string& confi
+
+     g_lua.registerFunction<void(DNSQuestion::*)(vector<pair<string, string>>)>("setTagArray", [](DNSQuestion& dq, const vector<pair<string, string>>&tags) {
+
+      setLuaSideEffect();
```
calls to setLuaSideEffect() and setLuaNoSideEffect() removed.

sethomisc commented on pdns/dnsdist-lua2.cc:
```diff
@@ -769,6 +769,59 @@ void moreLua(bool client)
+
+        message.setQueryTime(0, 0);                             // seconds and microseconds
+
+        message.addRRs(strQueryName);                           // add a RR to the response
```
New function setProtobufResponseRR() added.  setProtobufResponseType() does not add a RR.

sethomisc commented on pdns/dnsdist-lua2.cc:
```diff
@@ -769,6 +769,59 @@ void moreLua(bool client)
+     });
+
+                                                                // setProtobufResponseTypeQT - with query time as function parameter
+     g_lua.registerFunction<void(DNSDistProtoBufMessage::*)(const std::string&, time_t sec, uint uSec)>("setProtobufResponseTypeTS", [](DNSDistProtoBufMessage& message, const std::string& strQueryName, time_t sec, uint uSec) {
```
setProtobufResponseType() now can take optional parameters to set time.

sethomisc commented on pdns/dnsdist-lua2.cc:
```diff
@@ -769,6 +769,59 @@ void moreLua(bool client)
+
+        message.setType(DNSProtoBufMessage::Response);          // set protobuf type to be response - not query
+
+#ifdef TRASH
```
Removed #ifdef TRASH lines.

sethomisc commented:
Bert and Remi, 
     Thank you for your constructive comments.  I have attempted to do all of them.   In the structure DNSQuestion in dnsdist.hh I changed QTag to be a shared_ptr as I had a lot of difficulty in making it a unique_ptr.  Please let me know if any of my changes are unclear or need further work.

Thanks,

   Seth

rgacogne commented on pdns/dnsdist-lua.cc:
```diff
@@ -1572,6 +1572,57 @@ vector<std::function<void(void)>> setupLua(bool client, const std::string& confi
 
+    g_lua.registerFunction<void(DNSQuestion::*)(std::string, std::string)>("setTag", [](DNSQuestion& dq, const std::string& strLabel, const std::string& strValue) {
+
+      if(dq.qTag == NULL)
```
It would be a bit nicer to use `nullptr` instead of `NULL`, as the former can't be confused with an integer.

rgacogne commented on pdns/dnsdist-lua.cc:
```diff
@@ -1572,6 +1572,57 @@ vector<std::function<void(void)>> setupLua(bool client, const std::string& confi
+    g_lua.registerFunction<void(DNSQuestion::*)(std::string, std::string)>("setTag", [](DNSQuestion& dq, const std::string& strLabel, const std::string& strValue) {
+
+      if(dq.qTag == NULL)
+        {
```
We usually place the `{` following an `if` or `else` on the same line.

rgacogne commented on pdns/dnsdist-lua.cc:
```diff
@@ -1572,6 +1572,57 @@ vector<std::function<void(void)>> setupLua(bool client, const std::string& confi
+
+      if(dq.qTag == NULL)
+        {
+         dq.qTag = std::shared_ptr<QTag>(new QTag);
```
Replacing `std::shared_ptr<QTag>(new QTag)` by `std::make_shared<QTag>()` would save an allocation and be a bit more easy to read.

rgacogne commented on pdns/dnsdist-lua.cc:
```diff
@@ -1572,6 +1572,57 @@ vector<std::function<void(void)>> setupLua(bool client, const std::string& confi
+      if(dq.qTag == NULL)
+        {
+         dq.qTag = std::shared_ptr<QTag>(new QTag);
+        }
```
We usually don't indent the closing `}`

rgacogne commented on pdns/dnsdist-lua2.cc:
```diff
@@ -769,6 +769,47 @@ void moreLua(bool client)
+
+     });
+
+     g_lua.registerFunction<void(DNSDistProtoBufMessage::*)(const std::string&, uint uType, uint uClass, uint uTTL, vector<pair<int, int>> )>("setProtobufResponseRR", [](DNSDistProtoBufMessage& message,
```
Please use `uint16_t` for type and class, `uint32_t` for ttl. A simple `std:string` might be more suited for `blobData`, it would prevent to have to allocate an array and copy the content to it?

rgacogne replied:
Perhaps `addResponseRR()` would be more suited, since we can add several RRs?

rgacogne commented on pdns/dnsdist.hh:
```diff
@@ -51,6 +54,95 @@ extern uint16_t g_ECSSourcePrefixV4;
+
+public:
+
+QTag()
```
Indentation looks wrong in the entire class.

rgacogne commented on pdns/dnsdist.hh:
```diff
@@ -51,6 +54,95 @@ extern uint16_t g_ECSSourcePrefixV4;
+bool bStatus = true;
+
+   tagData.insert( {strLabel, strValue});
+  return(bStatus);
```
It looks like it always return `true`, so perhaps the return value might be omitted?

rgacogne commented on pdns/dnsdist.hh:
```diff
@@ -51,6 +54,95 @@ extern uint16_t g_ECSSourcePrefixV4;
+    std::unordered_map<std::string, std::string>::const_iterator got =tagData.find (strLabel);
+    if(got == tagData.end())
+      {
+       return("");
```
We usually don't use parenthesis for `return` statements.

rgacogne commented on pdns/protobuf.cc:
```diff
@@ -97,6 +97,40 @@ void DNSProtoBufMessage::setEDNSSubnet(const Netmask& subnet, uint8_t mask)
+    return;
+
+  PBDNSMessage_DNSResponse_DNSRR* rr = response->add_rrs();
+      if (rr) {
```
Indentation looks off.

rgacogne replied:
still true

rgacogne commented on pdns/protobuf.cc:
```diff
@@ -97,6 +97,40 @@ void DNSProtoBufMessage::setEDNSSubnet(const Netmask& subnet, uint8_t mask)
+#endif /* HAVE_PROTOBUF */
+}
+
+void DNSProtoBufMessage::addRR(const std::string& strName, uint32_t uType, uint32_t uClass, uint32_t uTTL, const uint8_t *ptrBlob, size_t uBlobLen)
```
`uType` and `uClass` should be `uint16_t`, `ptrBlob` and `uBlobLen` can be merged into a `std::string`.

rgacogne commented on pdns/dnsdist-lua.cc:
```diff
@@ -1572,6 +1572,57 @@ vector<std::function<void(void)>> setupLua(bool client, const std::string& confi
+     });
+
+
+     g_lua.registerFunction<string(DNSQuestion::*)(std::string)>("getTagMatch", [](const DNSQuestion& dq, const std::string& strLabel) {
```
Perhaps `getTag()` would be more logical, to match `setTag()`?

sethomisc commented:
Remi,
Thanks for your very constructive comments, I've gone through the list and corrected them.
I'm not clear on how I should document my changes, any direction on that would also be appreciated.
I've attached a very brief description of the new lua commands with examples below.
Seth

[ReadMe-7-5-2017.txt](https://github.com/PowerDNS/pdns/files/1128440/ReadMe-7-5-2017.txt)

pieterlexis commented:
@sethomisc I have created new documentation for the upcoming 4.1 release (#5481), you can PR documentation there. Or if you want, I can add your readme contents to them.

rgacogne commented on pdns/dnsdist-lua2.cc:
```diff
@@ -769,6 +769,27 @@ void moreLua(bool client)
+     });
+
+    g_lua.registerFunction<void(DNSDistProtoBufMessage::*)(boost::optional <time_t> sec, boost::optional <uint> uSec)>("setProtobufResponseType",
+                                        [](DNSDistProtoBufMessage& message, boost::optional <time_t> sec, boost::optional <uint> uSec) {
```
`uSec` should be a `uint32_t` here to be consistent with `DNSProtoBufMessage::setQueryTime()`. Be careful that you need to change it twice due to the `LuaWrapper` syntax.

rgacogne commented on pdns/dnsdist-lua2.cc:
```diff
@@ -769,6 +769,27 @@ void moreLua(bool client)
+     });
+
+    g_lua.registerFunction<void(DNSDistProtoBufMessage::*)(const std::string& strQueryName, uint uType, uint uClass, uint uTTL, const std::string& strBlob)>("addResponseRR", [](DNSDistProtoBufMessage& message,
+                                                            const std::string& strQueryName, uint16_t uType, uint uClass, uint32_t uTTL, const std::string& strBlob) {
```
`uType` should be `uint16_t` on the first line too. `uClass` should be `uint16_t` on both lines

rgacogne commented on pdns/dnsdist.hh:
```diff
@@ -51,6 +54,87 @@ extern uint16_t g_ECSSourcePrefixV4;
+std::string getEntry(size_t iEntry) const
+{
+std::string strEntry;
+size_t iCounter = 0;
```
small indentation issue

rgacogne commented on pdns/protobuf.cc:
```diff
@@ -97,6 +97,39 @@ void DNSProtoBufMessage::setEDNSSubnet(const Netmask& subnet, uint8_t mask)
+#endif /* HAVE_PROTOBUF */
+}
+
+void DNSProtoBufMessage::addRR(const std::string& strName, uint32_t uType, uint32_t uClass, uint32_t uTTL, const std::string& strBlob)
```
`uType` and `uClass` should be `uint16_t`.

rgacogne commented on regression-tests.dnsdist/test_Protobuf.py:
```diff
@@ -10,32 +10,110 @@
-        requestor = newCA(dq.remoteaddr:toString())
-        if requestor:isIPv4() then
-          requestor:truncate(24)
+    luasmn = newSuffixMatchNode()				
```
You have inserted a lot of trailing spaces, could you fix that?

rgacogne commented on regression-tests.dnsdist/test_Protobuf.py:
```diff
@@ -94,22 +173,28 @@ def checkProtobufBase(self, msg, protocol, query, initiator):
+		self.assertEquals(msg.type, dnsmessage_pb2.PBDNSMessage.DNSQueryType)		# testProtobuf()
+	else:
+		self.assertEquals(msg.type, dnsmessage_pb2.PBDNSMessage.DNSResponseType)	# testLuaProtobuf()
+
```
That blocks looks very weird because it's effectively making `checkProtobufQuery()` accept responses. Perhaps we should call `checkProtobufResponse()` when the query has been turned into a response instead?

rgacogne commented on pdns/dnsdist.hh:
```diff
@@ -51,6 +54,82 @@ extern uint16_t g_ECSSourcePrefixV4;
+{
+}
+
+void add(std::string strLabel, std::string strValue)
```
`strLabel` and `strValue` could be `const` references instead

rgacogne commented on pdns/dnsdist.hh:
```diff
@@ -51,6 +54,82 @@ extern uint16_t g_ECSSourcePrefixV4;
+  std::unordered_map<std::string, std::string>tagData;
+
+private:
+  const char *strSep = "\t";
```
Could be `static`

rgacogne commented on pdns/protobuf.cc:
```diff
@@ -97,6 +97,38 @@ void DNSProtoBufMessage::setEDNSSubnet(const Netmask& subnet, uint8_t mask)
+#endif /* HAVE_PROTOBUF */
+}
+
+void DNSProtoBufMessage::addRR(const std::string& strName, uint16_t uType, uint16_t uClass, uint32_t uTTL, const std::string& strBlob)
```
`strName` should probably be a `const DNSName&` instead of a string

rgacogne commented on pdns/protobuf.cc:
```diff
@@ -97,6 +97,38 @@ void DNSProtoBufMessage::setEDNSSubnet(const Netmask& subnet, uint8_t mask)
+    rr->set_type(uType);
+    rr->set_class_(uClass);
+    rr->set_ttl(uTTL);
+    rr->set_rdata((const uint8_t *) strBlob.c_str(), strBlob.size());
```
If I'm not mistaken the C-style cast to `const uint8_t *` is not needed.

sethomisc commented:
Remi,

   changed the strLabel and strValue to be const references in Qtag::add() located in dnsdist.hh

   changed const char *strSep =\t";  to static constexpr char const *strSep = "\t"; in dnsdist.hh

   removed the (const uint8 t *) cast in DNSProBufMessage:addRR() in protobuf.cc

  unsure if I should change strName from a string to a const DNSName& as I intended to call the function with just the DNS string like dr.qname:toString(), please advise me Remi.

Again thank you for all your input Remi,   Seth

sethomisc commented:
Remi, 
    I have rebased dnsdist-mod2.
    Please let me know if there are any problems.
    Seth

rgacogne commented:
Hi Seth! Thanks, there is an issue with your last commit that completely removes the `README-dnsdist.md` file, causing the test to file. Could you just drop that last commit and amend e75a021ff1c2426589e8ac9f3a428b3931d2fbcd ("Fixed requests from Remi July 7") to remove any alteration to that file?

nesv commented:
@rgacogne Along with your requested amendment to e75a021, I have also removed some other files that were introduced by that commit, that probably shouldn't have been there:

- `pdns/dnsdistdist/cdb.so`
- `contrib/dnsmessage_pb2.py{,c}`

rgacogne commented:
Merged, thank you again!
